### PR TITLE
nvproxy: implement missing NV_MEMORY_MULTICAST_FABRIC, NV00FD_*

### PR DESCRIPTION
**Pull request by thundergolfer**
Adding ioctls to fix a simple multi-GPU Huggingface`accelerate` program that does not work on GCP H100s.

---

### System details

* **instance type:** `a3-highgpu-8g` (GCP, us-east4-a)
* **NVIDIA driver:** `Driver Version: 550.54.15      CUDA Version: 12.4` 
* **NVIDIA device:**  4 x NVIDIA H100 HBM3
* **uname -a:** `Linux gcp-h100-us-east4-a-0-bb25baf985414f8899dfdfcb82d6796d 5.15.0-208.159.3.el9uek.x86_64 #2 SMP Wed Jun 19 09:05:13 PDT 2024 x86_64 x86_64 x86_64 GNU/Linux` 

```
runsc version release-20240513.0-173-gc526d251933a-dirty
spec: 1.1.0-rc.1
```

---

## Reproduction steps

1. **Install gVisor**

```bash
ARCH="x86_64"  # all workers are x86
URL="https://storage.googleapis.com/gvisor/releases/master/latest/${ARCH}"
wget -nc "${URL}/runsc" "${URL}/runsc.sha512"
chmod +x runsc
sudo cp runsc /usr/local/bin/runsc
sudo /usr/local/bin/runsc install
sudo systemctl reload docker
```

**2. Add GPU enabling gvisor options**

In `/etc/docker/daemon.json`:

```json
{
    "runtimes": {
        "nvidia": {
            "path": "nvidia-container-runtime",
            "runtimeArgs": []
        },
        "runsc": {
            "path": "/home/modal/runsc",
	    "runtimeArgs": ["--nvproxy", "--nvproxy-docker", "-debug-log=/tmp/runsc/", "-debug", "-strace"]

        }
    }
}
```

**3. Run Dockerfile**

```Dockerfile
# Dockerfile
FROM winglian/axolotl@sha256:5c724f7accd8188b0f84ead93b7efbfa8f8661f40e133646bd6d946bc3423d6d
RUN pip install fastapi==0.111.0
RUN pip install huggingface-hub~=0.23.0 pydantic==2.6.3 python-dateutil
ENV HUGGINGFACE_HUB_CACHE="/pretrained"
ENV TQDM_DISABLE="true"
ENV AXOLOTL_NCCL_TIMEOUT="60"

COPY <<EOF repro.py
import os
import subprocess
from pathlib import Path

print("[MOD-3226] hello from the repro!!!")
from accelerate import Accelerator

accelerator = Accelerator()

with accelerator.main_process_first():
    print(f"hello! {accelerator.process_index}")
EOF

ENTRYPOINT ["accelerate", "launch", "repro.py"]
```

```
sudo docker run -it --runtime=$RUNTIME --gpus='"device=GPU-c453e5c7-a56d-70bf-78ce-61be6cb8e0db,GPU-4703196a-e3df-9e3f-bb8b-6fa91c8e9970,GPU-4a9c162c-9280-eaa8-215a-2c681e82a99f,GPU-1660d344-e18b-e48a-cced-38380e903c31"' ce4326479c8412b13bba27416e3e77093d4411279b432ca1b25050f17ef57a67
```

### Results

**`runc`**

```
sudo docker run -it  --gpus='"device=GPU-c453e5c7-a56d-70bf-78ce-61be6cb8e0db,GPU-4703196a-e3df-9e3f-bb8b-6fa91c8e9970,GPU-4a9c162c-9280-eaa8-215a-2c681e82a99f,GPU-1660d344-e18b-e48a-cced-38380e903c31"' ce4326479c8412b13bba27416e3e77093d4411279b432ca1b25050f17ef57a67
The following values were not passed to `accelerate launch` and had defaults used instead:
	`--num_processes` was set to a value of `4`
		More than one GPU was found, enabling multi-GPU training.
		If this was unintended please pass in `--num_processes=1`.
	`--num_machines` was set to a value of `1`
	`--mixed_precision` was set to a value of `'no'`
	`--dynamo_backend` was set to a value of `'no'`
To avoid this warning pass in values for each of the problematic parameters or run `accelerate config`.
[MOD-3226] hello from the repro!!!
[MOD-3226] hello from the repro!!!
[MOD-3226] hello from the repro!!!
[MOD-3226] hello from the repro!!!
hello! 0
hello! 1
hello! 2hello! 3
```

**`runsc` (main)**

<details> <summary>💥 Failure logs</summary>

```
sudo docker run -it --runtime=runsc --gpus='"device=GPU-c453e5c7-a56d-70bf-78ce-61be6cb8e0db,GPU-4703196a-e3df-9e3f-bb8b-6fa91c8e9970,GPU-4a9c162c-9280-eaa8-215a-2c681e82a99f,GPU-1660d344-e18b-e48a-cced-38380e903c31"' ce4326479c8412b13bba27416e3e77093d4411279b432ca1b25050f17ef57a67
The following values were not passed to `accelerate launch` and had defaults used instead:
	`--num_processes` was set to a value of `4`
		More than one GPU was found, enabling multi-GPU training.
		If this was unintended please pass in `--num_processes=1`.
	`--num_machines` was set to a value of `1`
	`--mixed_precision` was set to a value of `'no'`
	`--dynamo_backend` was set to a value of `'no'`
To avoid this warning pass in values for each of the problematic parameters or run `accelerate config`.
[MOD-3226] hello from the repro!!!
[MOD-3226] hello from the repro!!!
[MOD-3226] hello from the repro!!!
[MOD-3226] hello from the repro!!!
Detected kernel version 4.4.0, which is below the recommended minimum of 5.5.0; this can cause the process to hang. It is recommended to upgrade the kernel to the minimum version or higher.
hello! 0
Traceback (most recent call last):
  File "/workspace/axolotl/repro.py", line 10, in <module>
    with accelerator.main_process_first():
  File "/root/miniconda3/envs/py3.10/lib/python3.10/contextlib.py", line 142, in __exit__
    next(self.gen)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/accelerator.py", line 884, in main_process_first
    with self.state.main_process_first():
  File "/root/miniconda3/envs/py3.10/lib/python3.10/contextlib.py", line 142, in __exit__
    next(self.gen)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/state.py", line 1056, in main_process_first
    with PartialState().main_process_first():
  File "/root/miniconda3/envs/py3.10/lib/python3.10/contextlib.py", line 142, in __exit__
    next(self.gen)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/state.py", line 502, in main_process_first
    yield from self._goes_first(self.is_main_process)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/state.py", line 390, in _goes_first
    self.wait_for_everyone()
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/state.py", line 379, in wait_for_everyone
    torch.distributed.barrier()
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/distributed/c10d_logger.py", line 47, in wrapper
    return func(*args, **kwargs)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/distributed/distributed_c10d.py", line 3696, in barrier
    work = default_pg.barrier(opts=opts)
torch.distributed.DistBackendError: NCCL error in: ../torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1333, unhandled cuda error (run with NCCL_DEBUG=INFO for details), NCCL version 2.18.6
ncclUnhandledCudaError: Call to CUDA function failed.
Last error:
Cuda failure 'unknown error'
[2024-07-11 19:52:01,530] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 68 closing signal SIGTERM
[2024-07-11 19:52:01,532] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 69 closing signal SIGTERM
[2024-07-11 19:52:01,533] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 70 closing signal SIGTERM
[2024-07-11 19:52:02,108] torch.distributed.elastic.multiprocessing.api: [ERROR] failed (exitcode: 1) local_rank: 0 (pid: 67) of binary: /root/miniconda3/envs/py3.10/bin/python
Traceback (most recent call last):
  File "/root/miniconda3/envs/py3.10/bin/accelerate", line 8, in <module>
    sys.exit(main())
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/commands/accelerate_cli.py", line 46, in main
    args.func(args)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/commands/launch.py", line 1073, in launch_command
    multi_gpu_launcher(args)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/commands/launch.py", line 718, in multi_gpu_launcher
    distrib_run.run(args)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/distributed/run.py", line 797, in run
    elastic_launch(
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 134, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 264, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError:
============================================================
repro.py FAILED
------------------------------------------------------------
Failures:
  <NO_OTHER_FAILURES>
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2024-07-11_19:52:01
  host      : d45a08528293
  rank      : 0 (local_rank: 0)
  exitcode  : 1 (pid: 67)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
============================================================
```

---

</details>

**`runsc` (this pull request)**

<details> <summary>✅ Success logs</summary>

```
[modal@gcp-h100-us-east4-a-0-bb25baf985414f8899dfdfcb82d6796d ~]$ sudo docker run -it --runtime=runsc --gpus='"device=GPU-c453e5c7-a56d-70bf-78ce-61be6cb8e0db,GPU-4703196a-e3df-9e3f-bb8b-6fa91c8e9970,GPU-4a9c162c-9280-eaa8-215a-2c681e82a99f,GPU-1660d344-e18b-e48a-cced-38380e903c31"' ce4326479c8412b13bba27416e3e77093d4411279b432ca1b25050f17ef57a67
The following values were not passed to `accelerate launch` and had defaults used instead:
	`--num_processes` was set to a value of `4`
		More than one GPU was found, enabling multi-GPU training.
		If this was unintended please pass in `--num_processes=1`.
	`--num_machines` was set to a value of `1`
	`--mixed_precision` was set to a value of `'no'`
	`--dynamo_backend` was set to a value of `'no'`
To avoid this warning pass in values for each of the problematic parameters or run `accelerate config`.
[MOD-3226] hello from the repro!!!
[MOD-3226] hello from the repro!!!
[MOD-3226] hello from the repro!!!
[MOD-3226] hello from the repro!!!
Detected kernel version 4.4.0, which is below the recommended minimum of 5.5.0; this can cause the process to hang. It is recommended to upgrade the kernel to the minimum version or higher.
hello! 0
hello! 1
hello! 3hello! 2
```

</details>

### CODE REVIEW

File: pkg/sentry/devices/nvproxy/frontend_unsafe.go [hunk 1]
```diff
@@ -47,6 +47,29 @@ func rmControlInvoke[Params any](fi *frontendIoctlState, ioctlParams *nvgpu.NVOS
...
+		return 0, linuxerr.EINVAL
+	}
+	ctrlParams.DevDescriptor = uint64(DevDesc.hostFD)
+	n, err := rmControlInvoke(fi, ioctlParams, ctrlParams)
+	ctrlParams.DevDescriptor = origDevDescriptor
+	if err != nil {
+		return n, err
+	}
+	// ctrlParams.DevDescriptor = origDevDescriptor
+	if _, err := ctrlParams.CopyOut(fi.t, addrFromP64(ioctlParams.Params)); err != nil {
```
Comment (AC-Dap) [inline pkg/sentry/devices/nvproxy/frontend_unsafe.go:67]: Do we technically need to copy out? Looking at the [driver code](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/caa2dd11a0db42e5d6130f627dd15ef404b7597d/src/common/sdk/nvidia/inc/ctrl/ctrl00fd.h#L153), it seems like all the fields are only used for input.

**Review by AC-Dap — CHANGES_REQUESTED** [review]

File: pkg/sentry/devices/nvproxy/frontend_unsafe.go [hunk 2]
```diff
@@ -47,6 +47,29 @@ func rmControlInvoke[Params any](fi *frontendIoctlState, ioctlParams *nvgpu.NVOS
 	return n, nil
 }
 
+func ctrlAttachGpuInvoke(fi *frontendIoctlState, ioctlParams *nvgpu.NVOS54Parameters, ctrlParams *nvgpu.NV00FD_CTRL_ATTACH_GPU_PARAMS) (uintptr, error) {
+	origDevDescriptor := ctrlParams.DevDescriptor
+	DevDescriptor, _ := fi.t.FDTable().Get(int32(origDevDescriptor))
+	if DevDescriptor == nil {
+		return 0, linuxerr.EINVAL
+	}
```
[the remark this answers is not in the document]
Comment (thundergolfer) [reply pkg/sentry/devices/nvproxy/frontend_unsafe.go:55]: Oh duh, thanks.

File: pkg/sentry/devices/nvproxy/frontend_unsafe.go (hunk 1 above)
Comment (thundergolfer) [reply pkg/sentry/devices/nvproxy/frontend_unsafe.go:67]: Oh right, so the command is all `[IN]` and won't read any data back, so we don't need to memcpy from host to guest.

Comment (thundergolfer) [reply pkg/sentry/devices/nvproxy/frontend_unsafe.go:67]: Works without this 👍

Comment (ayushr2) [reply pkg/sentry/devices/nvproxy/frontend_unsafe.go]: Good catch Anthony!

File: pkg/abi/nvgpu/classes.go [hunk 3]
```diff
@@ -333,6 +334,37 @@ type NV00F8_ALLOCATION_PARAMETERS struct {
...
+	Index      uint16
+	_          [6]byte
+	Alignment  uint64
+	AllocSize  uint64
+	PageSize   uint32
+	AllocFlags uint32
+	NumGPUs    uint32
+	_          uint32
+	POsEvent   P64
+}
```
Comment (ayushr2) [inline pkg/abi/nvgpu/classes.go:366]: This struct has changed 2 times in the driver version window we support: https://github.com/NVIDIA/open-gpu-kernel-modules/blame/caa2dd11a0db42e5d6130f627dd15ef404b7597d/src/common/sdk/nvidia/inc/class/cl00fd.h#L87

Once in `550.40.07`, but this change was just cosmetic. Two newlines were added in the struct definition.
And once in `545.23.06`, this was significant.  The first two fields (expPacket and index) were added. We will need to add versioning support for `545.23.06`.

File: pkg/sentry/devices/nvproxy/frontend.go [hunk 4]
```diff
@@ -564,6 +564,22 @@ func ctrlHasFrontendFD[Params any, PtrParams hasFrontendFDPtr[Params]](fi *front
...
+	}
+	if _, err := ctrlParams.CopyIn(fi.t, addrFromP64(ioctlParams.Params)); err != nil {
+		return 0, err
+	}
+
+	n, err := ctrlAttachGpuInvoke(fi, ioctlParams, &ctrlParams)
+	if err != nil {
+		return n, err
+	}
+	return n, nil
```
Comment (ayushr2) [inline pkg/sentry/devices/nvproxy/frontend.go:580]: return ctrlAttachGpuInvoke(fi, ioctlParams, &ctrlParams)

File: pkg/sentry/devices/nvproxy/frontend_unsafe.go [hunk 5]
```diff
@@ -47,6 +47,26 @@ func rmControlInvoke[Params any](fi *frontendIoctlState, ioctlParams *nvgpu.NVOS
 	return n, nil
 }
 
+func ctrlAttachGpuInvoke(fi *frontendIoctlState, ioctlParams *nvgpu.NVOS54Parameters, ctrlParams *nvgpu.NV00FD_CTRL_ATTACH_GPU_PARAMS) (uintptr, error) {
+	origDevDescriptor := ctrlParams.DevDescriptor
+	DevDescriptor, _ := fi.t.FDTable().Get(int32(origDevDescriptor))
```
Comment (ayushr2) [inline pkg/sentry/devices/nvproxy/frontend_unsafe.go:52]: In Golang, local variable names should not start with upper case. Same for `DevDesc` below.

File: pkg/sentry/devices/nvproxy/frontend_unsafe.go [hunk 6]
```diff
@@ -47,6 +47,26 @@ func rmControlInvoke[Params any](fi *frontendIoctlState, ioctlParams *nvgpu.NVOS
...
+	if !ok {
+		return 0, linuxerr.EINVAL
+	}
+	ctrlParams.DevDescriptor = uint64(DevDesc.hostFD)
+	n, err := rmControlInvoke(fi, ioctlParams, ctrlParams)
+	ctrlParams.DevDescriptor = origDevDescriptor
+	if err != nil {
+		return n, err
+	}
+	return n, nil
```
Comment (ayushr2) [inline pkg/sentry/devices/nvproxy/frontend_unsafe.go:67]: return n, err

File: pkg/sentry/devices/nvproxy/frontend_unsafe.go (hunk 5 above)
Comment (thundergolfer) [reply pkg/sentry/devices/nvproxy/frontend_unsafe.go:52]: Eugh damn, thanks, will fix. If there's a linter/fmter I could be running pre-push LMK.

Comment (EtiennePerot) [reply pkg/sentry/devices/nvproxy/frontend_unsafe.go:52]: You can run `make test TARGETS=pkg/sentry/devices/nvproxy/...` which will run the `nogo` tests which have some linting, but I'm not sure that covers this particular problem. It's just Go style: https://google.github.io/styleguide/go/decisions.html#variable-names

**Review by ayushr2 — APPROVED** [review]

File: pkg/sentry/devices/nvproxy/frontend_unsafe.go (hunk 5 above)
Comment (thundergolfer) [reply pkg/sentry/devices/nvproxy/frontend_unsafe.go]: Thanks for the info. Yeh I looked in CONTRIBUTING and saw there was a style guide I could follow but no automated tool. My eye isn't good at catching these golang lints yet.

File: pkg/abi/nvgpu/classes.go [hunk 7]
```diff
@@ -333,6 +334,49 @@ type NV00F8_ALLOCATION_PARAMETERS struct {
...
+
+// NV00FD_ALLOCATION_PARAMETERS is the alloc param type for NV_MEMORY_MULTICAST_FABRIC
+// from src/common/sdk/nvidia/inc/class/cl00fd.h
+//
+// +marshal
+type NV00FD_ALLOCATION_PARAMETERS struct {
+	ExpPacket  NV_EXPORT_MEM_PACKET
+	Index      uint16
+	_          [6]byte
```
Comment (nixprime) [inline pkg/abi/nvgpu/classes.go:358]: IIUC, the pre-545 version of this struct lacks these fields.

File: pkg/sentry/devices/nvproxy/frontend.go [hunk 8]
```diff
@@ -564,6 +564,18 @@ func ctrlHasFrontendFD[Params any, PtrParams hasFrontendFDPtr[Params]](fi *front
 	return n, nil
 }
 
+func ctrlAttachGpu(fi *frontendIoctlState, ioctlParams *nvgpu.NVOS54Parameters) (uintptr, error) {
```
Comment (nixprime) [inline pkg/sentry/devices/nvproxy/frontend.go:567]: Nit: `AttachGpu` is a bit broad; I think `ctrlMemoryMulticastFabricAttachGPU` would be more specific and consistent with other `ctrl*` functions. Same for the Invoke helper function.

File: pkg/abi/nvgpu/classes.go (hunk 7 above)
Comment (ayushr2) [reply pkg/abi/nvgpu/classes.go:358]: Ah good catch, I think `NV00FD_ALLOCATION_PARAMETERS_V545` aims to cover this fact but it wasn't removed from here after versioning support was added.

Comment (thundergolfer) [reply pkg/abi/nvgpu/classes.go:358]: Yep thanks for catching!

File: pkg/sentry/devices/nvproxy/frontend.go (hunk 8 above)
Comment (thundergolfer) [reply pkg/sentry/devices/nvproxy/frontend.go]: Agree, thanks. Fixed in push.

**Review by AC-Dap — APPROVED** [review]

**Pull request merged**